src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt

```diff
@@ -5,26 +5,224 @@ import org.slf4j.LoggerFactory
+
+    val sessionTemplateEntity = SessionTemplate(
+      prisonId = prison.id,
+      prison = prison,
+      name = createSessionTemplateDto.name,
+      startTime = createSessionTemplateDto.startTime,
+      endTime = createSessionTemplateDto.endTime,
+      validFromDate = createSessionTemplateDto.validFromDate,
+      validToDate = createSessionTemplateDto.validToDate,
+      visitRoom = createSessionTemplateDto.visitRoom,
+      closedCapacity = createSessionTemplateDto.closedCapacity,
+      openCapacity = createSessionTemplateDto.openCapacity,
+      biWeekly = createSessionTemplateDto.biWeekly,
+      enhanced = createSessionTemplateDto.enhanced,
+      dayOfWeek = createSessionTemplateDto.dayOfWeek,
+      visitType = VisitType.SOCIAL
+    )
+
+    createSessionTemplateDto.locationGroupReferences?.let {
+      it.forEach { ref -> sessionTemplateEntity.permittedSessionGroups.add(this.getLocationGroupByReference(ref)) }
+    }
+
+    val sessionTemplateEntitySave = sessionTemplateRepository.saveAndFlush(sessionTemplateEntity)
+
+    return SessionTemplateDto(
+      sessionTemplateEntitySave
+    )
+  }
+
+  fun updateSessionTemplate(reference: String, updateSessionTemplateDto: UpdateSessionTemplateDto): SessionTemplateDto {
+
+    with(updateSessionTemplateDto) {
+      name?.let {
+        sessionTemplateRepository.updateNameByReference(reference, name)
+      }
+
+      startTime?.let {
+        sessionTemplateRepository.updateStartTimeByReference(reference, startTime)
+      }
+
+      endTime?.let {
+        sessionTemplateRepository.updateEndTimeByReference(reference, endTime)
+      }
+
+      validFromDate?.let {
+        sessionTemplateRepository.updateValidFromDateByReference(reference, validFromDate)
+      }
+
+      validToDate?.let {
+        sessionTemplateRepository.updateValidToDateByReference(reference, validToDate)
+      }
+
+      closedCapacity?.let {
+        sessionTemplateRepository.updateClosedCapacityByReference(reference, closedCapacity)
+      }
+
+      openCapacity?.let {
+        sessionTemplateRepository.updateOpenCapacityByReference(reference, openCapacity)
+      }
+
+      enhanced?.let {
+        sessionTemplateRepository.updateEnhancedByReference(reference, enhanced)
+      }
+
+      biWeekly?.let {
+        sessionTemplateRepository.updateBiWeeklyByReference(reference, biWeekly)
+      }
+    }
+
+    val updatedSessionTemplateEntity = getSessionTemplate(reference)
+
+    updateSessionTemplateDto.locationGroupReferences?.let {
+      it.forEach { ref -> updatedSessionTemplateEntity.permittedSessionGroups.add(this.getLocationGroupByReference(ref)) }
+    }
+
+    return SessionTemplateDto(updatedSessionTemplateEntity)
+  }
+
+  fun deleteSessionTemplates(reference: String) {
+    val deleted = sessionTemplateRepository.deleteByReference(reference)
+    if (deleted == 0) {
+      throw ItemNotFoundException("Session template not found : $reference")
+    }
+    if (deleted > 1) {
```

Comment:
Would this be a very unlikely scenario?

Reply:
True, should I take it out! it's defensive programming I will take it out!